src/userobjects/RadialGreensConvolution.C

```diff
@@ -0,0 +1,419 @@
+  params.addCoupledVar("v", "Variable to gather");
+  params.addRequiredParam<FunctionName>("function",
+                                        "Green's function (distance is substituted for x) without geometrical attenuation");
+  params.addRequiredParam<Real>("r_cut", "Cut-off radius for the Green's function");
+  params.addParam<bool>("normalize", false, "Normalize the Green's function integral to one");
+
+  // we run this object once at the beginning of the timestep by default
+  params.set<ExecFlagEnum>("execute_on") = EXEC_TIMESTEP_BEGIN;
+
+  return params;
+}
+
+RadialGreensConvolution::RadialGreensConvolution(const InputParameters & parameters)
+  : ElementUserObject(parameters),
+    _v(coupledValue("v")),
+    _v_var(coupled("v")),
+    _function(getFunction("function")),
+    _r_cut(getParam<Real>("r_cut")),
+    _normalize(getParam<bool>("normalize")),
+    _dim(_mesh.dimension()),
+    _correction_integral(100)
+{
+  // collect mesh periodicity data
+  for (unsigned int i = 0; i < _dim; ++i)
+  {
+    _periodic[i] = _mesh.isRegularOrthogonal() && _mesh.isTranslatedPeriodic(_v_var, i);
+
+    _periodic_min[i] = _mesh.getMinInDimension(i);
+    _periodic_max[i] = _mesh.getMaxInDimension(i);
+    _periodic_vector[i](i) = _mesh.dimensionWidth(i);
+
+    // we could allow this, but then we'd have to search over more than just the nearest periodic
+    // neighbors
+    if (_periodic[i] && 2.0 * _r_cut > _periodic_vector[i](i))
+      paramError(
+          "r_cut",
+          "The cut-off radius cannot be larger than half the periodic size of the simulation cell");
+  }
+}
+
+void
+RadialGreensConvolution::initialize()
+{
+  _qp_data.clear();
+
+  /* compute the dimensional correction
+   *
+   *  |---__
+   *  |     / .
+   *  |  R/  | .
+   *  | /   h|  \ r_cut
+   *  |______|__|____
+   *         r
+   *
+   * Out of plane (2D) and out of line (1D) 3D contributions to the convolution
+   * are not naturally captured by the 1D or 2D numerical integration of QP neighborhood.
+   * At every Qp an integration in the h direction needs to be performed.
+   * We can pre-tabulate this integral.
+   */
+  if (_dim < 3)
+  {
+    const Real dr = _r_cut / _correction_integral.size();
+    for (unsigned int i = 0; i < _correction_integral.size(); ++i)
+    {
+      const Real r = i * dr;
+      const Real h = std::sqrt(_r_cut * _r_cut - r * r);
+      const unsigned int hsteps = std::ceil(h/dr);
+      const Real dh = h / hsteps;
+
+      _correction_integral[i] = 0.0;
+
+      // for r = 0 we skip the first bin of the integral!
+      if (i == 0)
+        _zero_dh = dh;
+
+      for (unsigned int j = i > 0 ? 0 : 1; j < hsteps; ++j)
+      {
+        const Real h1 = j * dh;
+        const Real h2 = h1 + dh;
+
+        // we evaluate the Green's function at the geometric middle of the height interval
+        // and assume it to be constant.
+        const Real R2 = r * r + h1 * h2;
```

Comment:
why not arithmetic middle?

Reply:
because it is shorter than writing `(h1 + h2) * (h1 + h2) / 4.0`